// @capacitor/assets/dist/project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Project = void 0;
const project_1 = require("@trapezedev/project");
const path_1 = require("path");
const utils_fs_1 = require("@ionic/utils-fs");
const input_asset_1 = require("./input-asset");
const log_1 = require("./util/log");
class Project extends project_1.MobileProject {
    constructor(projectRoot = process.cwd(), config, assetPath = 'assets') {
        super(projectRoot, config);
        this.assetPath = assetPath;
        this.assets = null;
        this.directory = null;
        // const projectRoot = join((config.android ?? config.ios)?.path ?? '', '../');
        this.directory = projectRoot;
        this.assetDir = (0, path_1.join)(projectRoot, assetPath);
        this.detectAssetDir(projectRoot);
    }
    async detectAssetDir(projectRoot) {
        if (this.assetPath === 'assets' && !(await (0, utils_fs_1.pathExists)(this.assetDir))) {
            this.assetDir = (0, path_1.join)(projectRoot, 'resources');
        }
        if (!(await (0, utils_fs_1.pathExists)(this.assetDir))) {
            (0, log_1.error)('Unable to find asset directory. Specify the --assetPath to the directory containing your assets, or create an assets folder in the top of your project');
        }
    }
    assetDirectory() {
        return this.assetDir;
    }
    async loadInputAssets() {
        this.assets = {
            logo: await this.loadInputAsset('logo', "logo" /* Logo */),
            logoDark: await this.loadInputAsset('logo-dark', "logo-dark" /* LogoDark */),
            icon: await this.loadInputAsset('icon', "icon" /* Icon */),
            iconForeground: await this.loadInputAsset('icon-foreground', "icon-foreground" /* IconForeground */),
            iconBackground: await this.loadInputAsset('icon-background', "icon-background" /* IconBackground */),
            splash: await this.loadInputAsset('splash', "splash" /* Splash */),
            splashDark: await this.loadInputAsset('splash-dark', "splash-dark" /* SplashDark */),
            iosIcon: await this.loadInputAsset('ios/icon', "icon" /* Icon */),
            iosSplash: await this.loadInputAsset('ios/splash', "splash" /* Splash */),
            iosSplashDark: await this.loadInputAsset('ios/splash-dark', "splash-dark" /* SplashDark */),
            iosNotificationIcon: await this.loadInputAsset('ios/notification-icon', "notification-icon" /* NotificationIcon */),
            iosSettingsIcon: await this.loadInputAsset('ios/settings-icon', "settings-icon" /* SettingsIcon */),
            iosSpotlightIcon: await this.loadInputAsset('ios/spotlight-icon', "spotlight-icon" /* SpotlightIcon */),
            androidIcon: await this.loadInputAsset('android/icon', "icon" /* Icon */),
            androidIconForeground: await this.loadInputAsset('android/icon-foreground', "icon" /* Icon */),
            androidIconBackground: await this.loadInputAsset('android/icon-background', "icon" /* Icon */),
            androidSplash: await this.loadInputAsset('android/splash', "splash" /* Splash */),
            androidSplashDark: await this.loadInputAsset('android/splash-dark', "splash-dark" /* SplashDark */),
            androidNotificationIcon: await this.loadInputAsset('android/notification', "notification-icon" /* NotificationIcon */),
        };
        return this.assets;
    }
    async loadInputAsset(path, kind) {
        let imagePath = null;
        const extensions = ['.png', '.webp', '.jpg', '.jpeg', '.svg'];
        let filename = null;
        for (let ext of extensions) {
            filename = `${path}${ext}`;
            if (await (0, utils_fs_1.pathExists)((0, path_1.join)(this.assetDir, filename))) {
                imagePath = (0, path_1.join)(this.assetDir, filename);
                break;
            }
        }
        if (!imagePath) {
            return null;
        }
        const asset = new input_asset_1.InputAsset(imagePath, kind);
        try {
            await asset.load();
            return asset;
        }
        catch (e) {
            (0, log_1.error)(`Unable to load source image ${filename}: ${e.message}`);
            return null;
        }
    }
}
exports.Project = Project;
